### ISSUE TITLE
PointRend: no 'pred_masks' field in instances when there are no predictions

### ISSUE THREAD

**Issue by MiroPsota** [facebookresearch/detectron2#2975]
[This commit](https://github.com/facebookresearch/detectron2/commit/c1689c7d7b1c1af0157cd90b84f296b7b2ce2124) changes behavior of `PointRendMaskHead` when there are no predictions on image. It does not add field `pred_masks` as before this commit. I expect this is not wanted behavior, as MaskRCNN also adds this field.

I think it's because there's no call `mask_rcnn_inference` now, when there are no predictions.
[before](https://github.com/facebookresearch/detectron2/blob/1dac147f824a3c8d908ff253cb8027f15844eb2f/projects/PointRend/point_rend/mask_head.py#L197)
[now](https://github.com/facebookresearch/detectron2/blob/master/projects/PointRend/point_rend/mask_head.py#L307)

I can provide MWE if needed.

**Comment by ppwwyyxx** [comment]
@bowenc0221 any idea why
```
        if len(pred_classes) == 0:
            return instances
```
was added?

**Comment by bowenc0221** [comment]
```
        if len(pred_classes) == 0:
            return instances
```
was added to make it compatible with Implicit PointRend. If there is no instance predicted, the parameter head will be an empty tensor and cause an error when doing convolution.

Any idea to better resolve this?

**Comment by ppwwyyxx** [comment]
torch's convolution ops are able to deal with tensors with batch size 0 (which is why this can work in mask r-cnn). do you have more specific examples?

**Comment by bowenc0221** [comment]
Implicit PointRend takes this batch size 0 tensor and predicts the `weight` of some group convolution which will fail? Because the group size = # of instance which is 0.

And because Implicit PointRend does not predict `coarse_mask`, we do not have this `mask_logits` so I removed `mask_rcnn_inference`.

We can fix it with something like this?
```python
        if len(pred_classes) == 0:
            return mask_rcnn_inference(torch.zeros((0, 28, 28)), instances)
```

**Comment by MiroPsota** [comment]
This seems to work:
```
        if len(pred_classes) == 0:
            mask_rcnn_inference(torch.zeros((0, 1, 28, 28)), instances)
            return instances
```

**Comment by ppwwyyxx** [comment]
It would be good to make sure all components work with empty inputs - they are so common in detection models. For the case in implicit pointrend, replacing 1x1 conv with a matmul will support empty size and also reads more clear to me:
```python
import torch
from torch.nn import functional as F

num_instances = 0
C = 8
Cout = 16
num_points = 10

features = torch.rand(1, num_instances * C, num_points)
weights = torch.rand(num_instances * Cout, C, 1)
bias = torch.rand(num_instances * Cout)

out1 = F.conv1d(features, weights, bias, stride=1, padding=0,
        groups=num_instances) # fails for N=0

features = features.reshape(num_instances, C, num_points)
weights = weights.reshape(num_instances, Cout, C)
bias = bias.reshape(num_instances, Cout, 1)
out2 = torch.einsum("nck,ndc->ndk", features, weights) + bias

print(out2.shape)
print(torch.abs(out1.reshape(out2.shape)-out2).max())
```

**Comment by ppwwyyxx** [comment]
Should be fixed by #2998
